pip/utils/__init__.py

```diff
@@ -105,28 +105,31 @@ def find_command(cmd, paths=None, pathext=None):
     # check if there are funny path extensions for executables, e.g. Windows
     if pathext is None:
         pathext = get_pathext()
-    pathext = [ext for ext in pathext.lower().split(os.pathsep) if len(ext)]
+    pathext = [ext for ext in pathext.lower().split(os.pathsep) if ext]
     # don't use extensions if the command ends with one of them
     if os.path.splitext(cmd)[1].lower() in pathext:
-        pathext = ['']
+        pathext = []
     # check if we find the command on PATH
     for path in paths:
-        # try without extension first
         cmd_path = os.path.join(path, cmd)
+        # if we have extensions, try them
         for ext in pathext:
-            # then including the extension
             cmd_path_ext = cmd_path + ext
             if os.path.isfile(cmd_path_ext):
                 return cmd_path_ext
-        if os.path.isfile(cmd_path):
-            return cmd_path
+        # else, don't use any extension
+        if not pathext:
+            if os.path.isfile(cmd_path):
+                return cmd_path
     raise BadCommand('Cannot find command %r' % cmd)
```

Comment:
... oh, is this what does the Unix check? This is not at all clear, and I fear that the code here would be pretty much unmaintainable given that most people know either Unix or Windows, but not both, and the tests typically don't get run on both Windows and Unix for every change.

If you're going to change this function (as I've already said, I'd like to see a real-world issue to justify doing so) then I'd rather it was made *more* clear what the code paths for each OS were, rather than less.